src/stream/repeat_with.rs

```diff
@@ -0,0 +1,103 @@
+use std::marker::PhantomData;
+use std::pin::Pin;
+
+use crate::future::Future;
+use crate::stream::Stream;
+use crate::task::{Context, Poll};
+
+/// A stream that repeats elements of type `T` endlessly by applying a provided clousre.
+///
+/// This stream is constructed by the [`repeat_with`] function.
+///
+/// [`repeat_with`]: fn.repeat_with.html
+#[derive(Debug)]
+pub struct RepeatWith<F, Fut, A> {
+    f: F,
+    future: Option<Fut>,
+    __a: PhantomData<A>,
+}
+
+/// Creates a new stream that repeats elements of type `A` endlessly by applying the provided closure.
+///
+/// # Examples
+///
+/// Basic usage:
+///
+/// ```
+/// # fn main() { async_std::task::block_on(async {
+/// #
+/// use async_std::prelude::*;
+/// use async_std::stream;
+///
+/// let s = stream::repeat_with(|| async { 1 });
+///
+///
+/// pin_utils::pin_mut!(s);
+///
+/// assert_eq!(s.next().await, Some(1));
+/// assert_eq!(s.next().await, Some(1));
+/// assert_eq!(s.next().await, Some(1));
+/// assert_eq!(s.next().await, Some(1));
+///
+/// # }) }
+/// ```
+///
+/// Going finite:
+///
+/// ```
+/// # fn main() { async_std::task::block_on(async {
+/// #
+/// use async_std::prelude::*;
+/// use async_std::stream;
+///
+/// let s = stream::repeat_with(|| async { 1u8 }).take(2);
+///
+///
+/// pin_utils::pin_mut!(s);
+///
+/// assert_eq!(s.next().await, Some(1));
+/// assert_eq!(s.next().await, Some(1));
+/// assert_eq!(s.next().await, None);
+///
+/// # }) }
+/// ```
+pub fn repeat_with<F, Fut, A>(repeater: F) -> RepeatWith<F, Fut, A> {
```

Comment:
In the [`std::iter::RepeatWith` docs](https://doc.rust-lang.org/std/iter/fn.repeat_with.html), the `RepeatWith` trait bounds are also included in the signature of the free function. I think this is nice because it's easier to figure out what the shape of the arguments should be when reading the docs. Could you perhaps add that?